distribution/pull_v2.go

```diff
@@ -40,14 +40,14 @@ var (
 	errRootFSInvalid  = errors.New("invalid rootfs in image configuration")
 )
 
-// ImageConfigPullError is an error pulling the image config blob
+// imageConfigPullError is an error pulling the image config blob
 // (only applies to schema2).
-type ImageConfigPullError struct {
+type imageConfigPullError struct {
```

Comment:
Tracing through the code, it looks like `distribution.Pull()` could return an error shaped like `errdefs.errUnknown{ImageConfigPullError{}}`. And since `errUnknown` has an `Unwrap()` method, callers could test for and retrieve the wrapped `ImageConfigPullError` value with `errors.As`...if the type is exported. Since errors of this type can escape the package, it should probably continue to be exported. And maybe it should grow an `Unwrap()` method, too?

Reply:
Hmm.. bit on the fence on the need to export it; looking at https://github.com/moby/moby/commit/8f26fe4f59ce515c68440da1443ace4c96e89d4a, it looks like it was purely created to show a "nicer" error, and it was never used outside of the package from the start.

Searching in (public) code on github;

- https://grep.app/search?q=ImageConfigPullError (used internally in docker code)
- https://grep.app/search?q=.ImageConfigPullError (no uses outside of the package)
- https://cs.github.com/?scopeName=All+repos&scope=&q=.ImageConfigPullError (no results either)